### ISSUE TITLE
EarlGrey 2.0 | grey_allOf fails with "Doesn't support pointer returns"

### ISSUE THREAD

**Issue by john-twigg-ck** [google/EarlGrey#1262]
Calling `grey_allOf` with any arguments fails over the bridge.

```
        let m = grey_accessibilityID("offers.preapproved")
        let matcher = grey_allOf([ m ]) // throws !!!!
```

https://github.com/google/eDistantObject/blob/c6d5422f98ada571a8bad727cee67c6845d2f44c/Service/Sources/EDOInvocationMessage.m#L434
```
else if (EDO_IS_POINTER(returnType)) {
          // TODO(haowoo): Handle this early and populate the exception.

          // We don't/can't support the plain memory access.
          NSAssert(NO, @"Doesn't support pointer returns."); <<<<BOOM!!!
        } 
```

```
/// "Doesn't support pointer returns"
 
#0	0x00007fff50b3b982 in objc_exception_throw ()
#1	0x00007fff23e3da48 in +[NSException raise:format:arguments:] ()
#2	0x00007fff258ac792 in -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] ()
#3	0x000000010a1d4265 in __38+[EDOInvocationRequest requestHandler]_block_invoke at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/Submodules/eDistantObject/Service/Sources/EDOInvocationMessage.m:434
#4	0x000000010a1dc314 in __51-[EDOHostService startReceivingRequestsForChannel:]_block_invoke.165 at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/Submodules/eDistantObject/Service/Sources/EDOHostService.m:445
#5	0x000000010a1e5cbc in -[EDOExecutorMessage executeBlock] at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/Submodules/eDistantObject/Service/Sources/EDOExecutorMessage.m:48
#6	0x000000010a1cee2b in -[EDOExecutor loopWithBlock:] at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/Submodules/eDistantObject/Service/Sources/EDOExecutor.m:83
#7	0x000000010a1e0529 in +[EDOClientService sendSynchronousRequest:onPort:withExecutor:] at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/Submodules/eDistantObject/Service/Sources/EDOClientService.m:273
#8	0x000000010a1ca8ae in -[EDOObject(Invocation) edo_forwardInvocation:selector:returnByValue:] at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/Submodules/eDistantObject/Service/Sources/EDOObject+Invocation.m:141
#9	0x000000010a1ca4d9 in -[EDOObject(Invocation) forwardInvocation:] at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/Submodules/eDistantObject/Service/Sources/EDOObject+Invocation.m:106
#10	0x00007fff23e42476 in ___forwarding___ ()
#11	0x00007fff23e449b8 in __forwarding_prep_0___ ()
#12	0x000000010a1adae1 in grey_allOfMatchers at /Users/christian.klug/git-repos/stable2/mob_ios/EarlGrey/AppFramework/Matcher/GREYMatchersShorthand.m:200

```

**Comment by john-twigg-ck** [comment]
It seems to fail the same way in your swift test `testButtonPressWithGREYAllOf` 

PS: I'm running Swift 5, XCode 11.4,  iOS 13.4

**Comment by john-twigg-ck** [comment]
Done some more digging. This may be a problem with `eDistantObject` 

The problem arrises in iOS 13.4.  Swift 5 and Xcode 11.4 are not the culprits I believe as iOS 13.3 seems to work.

The difference thats triggering the error is, for the method selector `objectAtIndex` , (obviously for the `NSArray`'s that play a role in this call), the Objc Runtime type is different than in previous versions...
```
char const *returnType = methodSignature.methodReturnType; 
// iOS 13.4:    ^@"
// iOS 13.3:    @
```

in iOS 13.4, this triggers `(EDO_IS_POINTER(returnType)` and `<Kaboom>`

**Comment by AlbertWang0116** [comment]
Hi John,

Thanks for looking into this. As you have figured out, as of iOS 13.4, Swift array declares its `objectAtIndex:` return type as "pointer to object" (`id*`) instead of object (`id`).

I'm surprised of this change, as Swift array's `objectAtIndex:` doesn't do what it declares now - it still returns you an object (`id`). It doesn't seem to be a global change to type encoding either, as the signature of my custom method is not changed.

I'm still looking at if there's any other side effects which are related to this. If you want to be unblocked right away, you can switch to NSArray when dealing with EarlGrey's API, such as grey_allOf. For example:

```
let matcherArray : NSArray = [grey_text("foo"), grey_sufficientlyVisible()]
let matcher = grey_allOf(matcherArray)
```

**Comment by wuhao5** [comment]
> `let matcher = grey_allOf([ m ]) // throws !!!!`

You can't really use Swift type as a remote object, because Swift does runtime checks in general. It shouldn't work and in Swift 5 it might have patched the hole to define Swift array more strict. The best way to fire a remote method is through @objc protocols, where Swift won't do the runtime check and goes thru the runtime dispatching machinery like what happens in ObjC.  So what you can do, as mentioned by @AlbertWang0116 , is 

```
grey_allOf(NSArray.init(array:[NSObject.init(), NSObject.init()]))
```

**Comment by john-twigg-ck** [comment]
@wuhao5 
`grey_allOf  is not a swift type.  This is not any functionality that I've added. It is an ObjC object `GREYAllOf`.

The issue exists in the latest branch of `earlgrey2` and I'm sure this issue will become pertinent as soon as this team tries to upgrade their targeting criteria to the latest iOS SDK of `13.4`. 

https://github.com/google/EarlGrey/blob/earlgrey2/AppFramework/Matcher/GREYMatchersShorthand.h#L180
```
GREY_EXPORT id<GREYMatcher> grey_allOfMatchers(NSArray<__kindof id<GREYMatcher>> *matchers)
    NS_SWIFT_NAME(grey_allOf(_:));
```

In the mean time, I've lowered my testing target to `iOS 13.3` to circumvent the issue.

**Comment by john-twigg-ck** [comment]
Here's my workaround solution for the time being, once which I happen to like anyways...

```
extension Sequence where Element == GREYMatcher {
    /// Use this instead of grey_allOf.
    /// Ex: [grey_accessibilityId(...), grey_buttonTitle(...)].all()
    func all() -> GREYMatcher {
        let m: [GREYMatcher] = self.map { $0 }
        // https://github.com/google/EarlGrey/issues/1262
        // swiftlint:disable:next force_cast
        return grey_allOf(NSArray(array: m) as! [GREYMatcher])
    }

    
    /// Use this instead of grey_anyOf.
    /// Ex: [grey_accessibilityId(...), grey_buttonTitle(...)].any()
    /// - Returns: <#description#>
    func any() -> GREYMatcher {
        let m: [GREYMatcher] = self.map { $0 }
        // https://github.com/google/EarlGrey/issues/1262
        // swiftlint:disable:next force_cast
        return grey_anyOf(NSArray(array: m) as! [GREYMatcher])
    }
}
```

**Comment by shaneong** [comment]
Any possibility of merging this as a permanent solution?

**Comment by AlbertWang0116** [comment]
This is fixed by EarlGrey 2.2.1 release/eDO 1.0.2 release so you don't need the workaround.
